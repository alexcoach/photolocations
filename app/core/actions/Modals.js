// @flow
export const MODALS_OPEN_NEXT = 'MODALS_OPEN_NEXT';
export const MODALS_CLOSE_CURRENT = 'MODALS_CLOSE_CURRENT';

import type { ModalConfig } from 'core/types';

export const openModal = (config: ModalConfig) => ({
    type: MODALS_OPEN_NEXT,
    config,
});

export const closeModal = () => ({
    type: MODALS_CLOSE_CURRENT,
});
