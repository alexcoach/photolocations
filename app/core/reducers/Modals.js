// @flow

import {
    MODALS_OPEN_NEXT,
    MODALS_CLOSE_CURRENT,
} from '../actions/Modals';

const DEFAULT_STATE = {
    queue: [],
};

const openNext = (state, { config }) => {
    const queue = state.queue;
    queue.push({
        id: Math.random().toString(),
        ...config,
    });
    return {
        ...state,
        queue,
    };
};

const closeCurrent = (state) => {
    const queue = state.queue;
    queue.pop();
    return {
        ...state,
        queue,
    };
};

export default (state:Object = DEFAULT_STATE, action:Object):Object => {
    switch (action.type) {
        case MODALS_OPEN_NEXT:
            return openNext(state, action);
        case MODALS_CLOSE_CURRENT:
            return closeCurrent(state);
        default:
            return state;
    }
};
