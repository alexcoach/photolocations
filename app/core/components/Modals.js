// @flow
import React from 'react';
import { connect } from 'react-redux';

import Modal from 'react-native-modal';
import { View } from 'native-base';

import type { ModalConfig } from 'core/types';

import Screen from './Screen';

import { closeModal } from '../actions/Modals';


const animations = {
    fade: {
        in: 'fadeIn',
        out: 'fadeOut',
    },
    zoom: {
        in: 'zoomIn',
        out: 'zoomOut',
    },
    slideUp: {
        in: 'slideInUp',
        out: 'slideOutDown',
    },
    slideDown: {
        in: 'slideInDown',
        out: 'slideOutUp',
    },
    slideRight: {
        in: 'slideInRight',
        out: 'slideOutLeft',
    },
    slideLeft: {
        in: 'slideInLeft',
        out: 'slideOutRight',
    },
};

const DEFAULT_ANIMATION = animations.fade;

type Props = {
    modalsQueueLength: number,
    modalsQueue: Array<ModalConfig>,
    dispatch: Function,
};

type State = {
    visibleModals: {
        [string]: number,
    },
};

class Modals extends React.PureComponent<Props, State> {
    state = {
        visibleModals: {},
    };
    componentWillReceiveProps(nextProps: Props) {
        if (this.props.modalsQueueLength !== nextProps.modalsQueueLength) {
            const visibleModals = {};
            nextProps.modalsQueue.forEach((modal) => {
                if (modal.id) {
                    visibleModals[modal.id] = 1;
                }
            });
            this.setState({
                visibleModals,
            });
        }
    }
    getModalStyle = (config: ModalConfig) => {
        const style = config.mode ? styles.modalMode[config.mode] : {};
        return {
            ...styles.modal,
            ...config.style,
            ...style,
        };
    };
    getModalAnimation = (config: ModalConfig): Object => {
        if ((typeof config.animation === 'string') && animations[config.animation]) {
            return animations[config.animation];
        }
        if (typeof config.animation === 'object') {
            return config.animation;
        }
        return DEFAULT_ANIMATION
    };
    closeCurrentModal = id => () => {
        const visibleModals = this.state.visibleModals;
        if (id && visibleModals[id] === 2) {
            visibleModals[id] = 0;
            this.setState({
                visibleModals,
            });
        }
    };
    unlockModal = id => () => {
        const visibleModals = this.state.visibleModals;
        if (id) {
            visibleModals[id] = 2;
            this.setState({
                visibleModals,
            });
        }
    };
    unregisterCurrentModal = () => this.props.dispatch(closeModal());
    renderModalContentWrap = (config: ModalConfig) => {
        if (config.screen) {
            const screenProps = {
                ...config.screen,
                left: { type: 'closeModal' },
            };
            return (
                <Screen {...screenProps}>
                    {config.content}
                </Screen>
            );
        }
        return (
            <View
                style={this.getModalStyle(config)}
            >
                {config.content}
            </View>
        );
    };
    renderModal = (config: ModalConfig, index: number) => {
        const modalConfig = config;
        if (modalConfig.screen) {
            modalConfig.mode = 'screen';
        }
        return (
            <View
                key={index}
                style={styles.modalWrap}
            >
                <Modal
                    style={styles.modalOverlay}
                    backdropOpacity={0.3}
                    animationIn={this.getModalAnimation(modalConfig).in}
                    animationOut={this.getModalAnimation(modalConfig).out}
                    isVisible={!modalConfig.id || !!this.state.visibleModals[modalConfig.id]}
                    onBackButtonPress={this.closeCurrentModal(modalConfig.id)}
                    onBackdropPress={this.closeCurrentModal(modalConfig.id)}
                    onModalHide={this.unregisterCurrentModal}
                    onModalShow={this.unlockModal(modalConfig.id)}
                    supportedOrientations={['portrait', 'landscape']}
                >
                    {this.renderModalContentWrap(modalConfig)}
                </Modal>
            </View>
        );
    };
    render() {
        const { modalsQueue } = this.props;
        return modalsQueue.map(this.renderModal);
    }
}

export default connect(state => ({
    modalsQueueLength: state.Modals.queue.length,
    modalsQueue: state.Modals.queue,
}))(Modals);

const styles = {
    modalOverlay: {
        flex: 1,
        alignItems: 'center',
        justifyContent: 'center',
        margin: 0,
    },
    modal: {
        minHeight: 100,
        minWidth: 300,
    },
    modalWrap: {
        position: 'absolute',
        top: 0,
        bottom: 0,
        left: 0,
        right: 0,
    },
    modalMode: {
        screen: {
            minWidth: '100%',
            minHeight: '100%',
        },
    },
};
