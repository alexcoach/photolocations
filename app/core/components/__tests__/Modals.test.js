import 'react-native';
import React from 'react';
import { shallow } from 'enzyme';
import configureStore from 'redux-mock-store';

import { Text } from 'native-base';

import Modals from '../Modals';

const mockStore = configureStore([]);

const state = {
    Modals: {
        queue: [
            {
                id: 'test',
                content: <Text>Modal Content</Text>
            }
        ],
    },
};

it('<Modals> renders correctly', () => {
    const wrapper = shallow(
        <Modals />,
        { context: { store: mockStore(state) } },
    );
    expect(wrapper.dive()).toMatchSnapshot();
});
