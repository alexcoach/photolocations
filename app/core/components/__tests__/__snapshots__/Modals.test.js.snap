// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Modals> renders correctly 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Modals
    dispatch={[Function]}
    modalsQueue={
      Array [
        Object {
          "content": <Styled(Text)>
            Modal Content
          </Styled(Text)>,
          "id": "test",
        },
      ]
    }
    modalsQueueLength={1}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Array [
    <Styled(ViewNB)
      style={
        Object {
          "bottom": 0,
          "left": 0,
          "position": "absolute",
          "right": 0,
          "top": 0,
        }
      }
    >
      <ReactNativeModal
        animationIn="fadeIn"
        animationInTiming={300}
        animationOut="fadeOut"
        animationOutTiming={300}
        avoidKeyboard={false}
        backdropColor="black"
        backdropOpacity={0.3}
        backdropTransitionInTiming={300}
        backdropTransitionOutTiming={300}
        hideModalContentWhileAnimating={false}
        isVisible={false}
        onBackButtonPress={[Function]}
        onBackdropPress={[Function]}
        onModalHide={[Function]}
        onModalShow={[Function]}
        scrollOffset={0}
        scrollOffsetMax={0}
        scrollTo={null}
        style={
          Object {
            "alignItems": "center",
            "flex": 1,
            "justifyContent": "center",
            "margin": 0,
          }
        }
        supportedOrientations={
          Array [
            "portrait",
            "landscape",
          ]
        }
        swipeThreshold={100}
        useNativeDriver={false}
      >
        <Styled(ViewNB)
          style={
            Object {
              "minHeight": 100,
              "minWidth": 300,
            }
          }
        >
          <Styled(Text)>
            Modal Content
          </Styled(Text)>
        </Styled(ViewNB)>
      </ReactNativeModal>
    </Styled(ViewNB)>,
  ],
  Symbol(enzyme.__nodes__): Array [
    Array [
      <Styled(ViewNB)
        style={
          Object {
            "bottom": 0,
            "left": 0,
            "position": "absolute",
            "right": 0,
            "top": 0,
          }
        }
      >
        <ReactNativeModal
          animationIn="fadeIn"
          animationInTiming={300}
          animationOut="fadeOut"
          animationOutTiming={300}
          avoidKeyboard={false}
          backdropColor="black"
          backdropOpacity={0.3}
          backdropTransitionInTiming={300}
          backdropTransitionOutTiming={300}
          hideModalContentWhileAnimating={false}
          isVisible={false}
          onBackButtonPress={[Function]}
          onBackdropPress={[Function]}
          onModalHide={[Function]}
          onModalShow={[Function]}
          scrollOffset={0}
          scrollOffsetMax={0}
          scrollTo={null}
          style={
            Object {
              "alignItems": "center",
              "flex": 1,
              "justifyContent": "center",
              "margin": 0,
            }
          }
          supportedOrientations={
            Array [
              "portrait",
              "landscape",
            ]
          }
          swipeThreshold={100}
          useNativeDriver={false}
        >
          <Styled(ViewNB)
            style={
              Object {
                "minHeight": 100,
                "minWidth": 300,
              }
            }
          >
            <Styled(Text)>
              Modal Content
            </Styled(Text)>
          </Styled(ViewNB)>
        </ReactNativeModal>
      </Styled(ViewNB)>,
    ],
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
    "context": Object {
      "store": Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      },
    },
  },
}
`;
